plugins/emoji/plugin.js

```diff
@@ -11,132 +11,96 @@
+			CKEDITOR.ajax.load( CKEDITOR.getUrl( emojiListUrl ), function( data ) {
+				if ( data === null ) {
 					return;
 				}
 
-				var filter = new CKEDITOR.htmlParser.filter( {
-					text: function( value, element ) {
-						var preventEmojiConversion = element.getAscendant( hasForbiddenParent );
-						if ( preventEmojiConversion ) {
-							return value;
-						}
-						var hits = value.match( /:[a-zA-Z_-]+?:/g );
-						if ( hits ) {
-							for ( var i = 0; i < hits.length; i++ ) {
-								var item = emoji.find( function( element ) {
-									return element.id === hits[ i ];
-								} );
-								if ( item ) {
-									value = value.replace( item.id, item.symbol );
-								}
-							}
-						}
-						return value;
-					}
-				} );
-				filter.applyTo( evt.data.dataValue );
-			} );
+				if ( editor._.emoji === undefined ) {
+					editor._.emoji = {};
+				}
 
-			function getTextTestCallback() {
-				return function( range ) {
-					if ( !range.collapsed || !isEmojiAllowed( range ) ) {
-						return null;
-					}
-					return CKEDITOR.plugins.textMatch.match( range, matchCallback );
-				};
-			}
+				if ( editor._.emoji.list === undefined ) {
+					editor._.emoji.list = JSON.parse( data );
+				}
 
-			function matchCallback( text, offset ) {
-				var left = text.slice( 0, offset ),
-					match = left.match( new RegExp( ':\\S{' + charactersToStart + '}\\S*$' ) );
+				var emojiList = editor._.emoji.list,
+					charactersToStart = editor.config.emoji_minChars === undefined ? 2 : editor.config.emoji_minChars;
 
-				if ( !match ) {
-					return null;
+				if ( editor.status !== 'ready' ) {
+					editor.once( 'instanceReady', function() {
+						initPlugin();
+					} );
+				} else {
+					initPlugin();
 				}
 
-				return { start: match.index, end: offset };
-			}
-
-			function dataCallback( query, range, callback ) {
-				var data = CKEDITOR.tools.array.filter( emoji, function( item ) {
-					return item.id.indexOf( query.slice( 1 ) ) !== -1;
-				} ).sort( function( a, b ) {
-					// Sort at the beginning emoji starts with given query.
-					var emojiName = query.substr( 1 ),
-						isAStartWithEmojiName = a.id.substr( 1, emojiName.length ) === emojiName,
-						isBStartWithEmojiName = b.id.substr( 1, emojiName.length ) === emojiName;
-
-					if ( isAStartWithEmojiName && isBStartWithEmojiName || !isAStartWithEmojiName && !isBStartWithEmojiName ) {
-						return a.id === b.id ? 0 : ( a.id > b.id ? 1 : -1 );
-					} else if ( isAStartWithEmojiName ) {
-						return -1;
-					} else {
-						return 1;
-					}
-				} );
-				callback( data );
-			}
+				// HELPER FUNCTIONS:
 
-			function isEmojiAllowed( range ) {
-				var elementsPath,
-					editable = editor.editable();
-				if ( range ) {
-					elementsPath = new CKEDITOR.dom.elementPath( range.startContainer, editable );
-					return elementsPath.contains( forbiddenScope ) ? false : true;
-				} else {
-					return true;
+				function initPlugin() {
```

Comment:
If we remove the `if` mentioned above, then this function could be inlined.

Reply:
related to previous comment :)